### PR TITLE
WIP: Edu apply wizard

### PR DESCRIPTION
**Pull request by cehsu**
Connects to department-of-veterans-affairs/vets.gov-team#1493

Description: The PR is intended more for soliciting feedback at this point rather than pulling into master, as it is a proposed change, and has not been formally approved/fully designed.

Instructions for testing this out, please: Navigate to `education/apply-wizard/`

Remaining TODOs
- [x] e2e testing
- [ ] User Testing to determine value of proposed change
- [ ] UI Feedback (May wish to include additional links straight to desired forms)

screen shot 2017-05-17 at 5 32 01 pm
__________________________________________________________________________________________________________
Subsequent questions are revealed once a selection has been made:
screen shot 2017-05-17 at 5 32 08 pm
__________________________________________________________________________________________________________
Visible questions are toggled once a parent selection has been updated:
screen shot 2017-05-17 at 5 32 15 pm

__________________________________________________________________________________________________________
Application button is revealed once the conclusion of a question path has been reached:
screen shot 2017-05-17 at 5 32 28 pm

### CODE REVIEW

File: src/js/no-react-entry.js
Comment (jbalboni) [inline src/js/no-react-entry.js]: This will keep the code from executing on non-edu pages, but it won't make a separate bundle. However, the file is only 65 lines, I'm not sure it's really worth it to split.

File: src/js/education-wizard.js
Comment (jbalboni) [inline src/js/education-wizard.js]: You shouldn't need these braces around the file.

File: src/js/education-wizard.js
Comment (jbalboni) [inline src/js/education-wizard.js]: You can use `Array.from()` instead of slice here, since this should now run after we load Babel's ES2105 polyfills.

File: src/js/education-wizard.js
Comment (jbalboni) [inline src/js/education-wizard.js]: This should probably be a change event. With click, this is going to run unnecessarily if someone clicks on an already selected option.

File: src/js/education-wizard.js
Comment (jbalboni) [inline src/js/education-wizard.js]: We probably also want to focus on the newly displayed question label.

**Comment by jbalboni** [comment]
@cehsu This looks great so far! One thing we likely need to do is make these changes in a copy of the apply page, so that we could push this to prod and make it easier for @goldenmeanie to user test, but not have it be live immediately.

Second, I noticed that if you fill out a few questions, then go back and change a previous answer, the subsequent questions/submit button don't always show up correctly. You might need to recalculate more/all of the wizard state on each input change.

**Comment by goldenmeanie** [comment]
I can run locally if the copy of the page is a pain

**Comment by jbalboni** [comment]
I mainly just don't want this PR to sit here for a couple weeks while we test things.

**Comment by goldenmeanie** [comment]
Whoa, it's real!

Some feedback:
- Questions need more vertical space between them (replicate in-app style)
- Additional questions should animate in (replicate in-app style)
- Eventual button should be "Apply Now" and width should be content driven, not full width

**Comment by cvalarida** [comment]
Hey, hey! Looks good. I'm pretty sure it's not actually connected to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2158, though. That issue is about refactoring the 1995 to use a shared page (which we created after we made the 1995).

File: src/sass/_shame-wizard.scss
Comment (cvalarida) [inline src/sass/_shame-wizard.scss]: Hehe, 'ware the _shame-wizard! He'll replay awkward teenage years again and again and again and...

Maybe rename this to `src/sass/modules/_m-education-wizard.scss`? It's perhaps not the _best_ place for it, but...

File: src/sass/_shame-wizard.scss
Comment (cvalarida) [inline src/sass/_shame-wizard.scss]: Do you need anything above this?

**Review by cvalarida — COMMENTED** [review]
I'll hold off on a thorough review until you're ready, but I had a couple things.

File: src/sass/_shame-wizard.scss
Comment (cehsu) [inline src/sass/_shame-wizard.scss]: I like this thought, I will just need to remember to clean up the unused styles in _shame.scss if this change gets approved.

File: src/sass/_shame-wizard.scss
Comment (cehsu) [inline src/sass/_shame-wizard.scss]: The original intention was for this to be a copy to swap in or out, but your idea to create a module is a better way to go about this.

**Comment by J-Quag** [comment]
I like where this is headed. :) Keep going! Looking forward to seeing the fleshed out version

File: src/js/education-wizard.js
Comment (jbalboni) [inline src/js/education-wizard.js]: Maybe we should put this file in `src/js/edu-benefits`? We don't have a folder for js for content pages aside from `legacy`, which seems misnamed.

File: src/js/no-react-entry.js
Comment (jbalboni) [inline src/js/no-react-entry.js]: I think you can leave this uncommented and do a build type check:

`if (__BUILDTYPE__ !== 'production' && location.href.endsWith('education/apply-wizard/')) {`

**Review by jbalboni — COMMENTED** [review]
Added some more comments. I think this is pretty close to ready. This is what I think is left:

- Clean up shame file
- Add build type check to no-react file
- Add /education/apply-wizard.md to the ignore list in build.js

File: src/js/edu-benefits/education-wizard.js
Comment (cvalarida) [inline src/js/edu-benefits/education-wizard.js]: Idle curiosity: @jbalboni Would your suggestion reduce the memory overhead by having only one event listener?

File: test/edu-benefits/edu-apply-wizard.e2e.spec.js
Comment (cvalarida) [inline test/edu-benefits/edu-apply-wizard.e2e.spec.js]: My guess is `/education/apply-wizard/` vs `/education/apply-wizard`, but :man_shrugging:

File: src/sass/_shame-wizard.scss
Comment (cvalarida) [inline src/sass/_shame-wizard.scss]: I just did a quick diff between _shame-wizard.scss and _shame.scss. The only difference is _shame.scss also has  the expander stuff used in the current apply page. We can just delete the shame wizard and stick with _shame. (We're currently not even using this file.)

File: src/sass/style.scss
Comment (cvalarida) [inline src/sass/style.scss]: Make like a cyberman and DELETE!

**Review by cvalarida — COMMENTED** [review]
Mmm...@goldenmeanie, should the apply button show under the warning, leading the user to the 1990e, or should it be hidden?
image

Like @jbalboni said, there's a little cleanup left, but it's looking good! I had a few questions, though.

File: src/js/edu-benefits/education-wizard.js
Comment (cehsu) [inline src/js/edu-benefits/education-wizard.js]: To satisfy: Require Consistent Returns (consistent-return)
https://github.com/department-of-veterans-affairs/vets-website/pull/5577

File: src/js/edu-benefits/education-wizard.js
Comment (cehsu) [inline src/js/edu-benefits/education-wizard.js]: To satisfy: Require Consistent Returns (consistent-return)
http://eslint.org/docs/2.0.0/rules/consistent-return

File: src/js/edu-benefits/education-wizard.js
Comment (cehsu) [inline src/js/edu-benefits/education-wizard.js]: @goldenmeanie explained they should still be able to apply, but that they should also be warned.

File: src/js/edu-benefits/education-wizard.js
Comment (cehsu) [inline src/js/edu-benefits/education-wizard.js]: I have change this so that it also checks if the warning is currently closed.

File: src/js/edu-benefits/education-wizard.js
Comment (cehsu) [inline src/js/edu-benefits/education-wizard.js]: I have changed this so that it also checks if the warning is currently closed.

**Comment by goldenmeanie** [comment]
@cvalarida I think that is a good question. My understanding is that this shouldn't necessarily stop people from applying, it's just that the application won’t go all the way through without the external step. So I think it makes sense to retain the button. Probably should adjust the copy though.

But that is a belief worth validating with stakeholders. I'll ask and get back to y’all

File: src/js/edu-benefits/education-wizard.js
Comment (cvalarida) [inline src/js/edu-benefits/education-wizard.js]: Oh, eslint, such a whiner. Alright, good enough for me. Alternatively, we could disable the rule with an [inline comment](http://eslint.org/docs/user-guide/configuring.html#disabling-rules-with-inline-comments), but eh. My curiosity is satisfied.

File: script/build.js
Comment (jbalboni) [inline script/build.js]: I think this needs to be `education/apply-wizard.md`

File: src/sass/_shame.scss
Comment (jbalboni) [inline src/sass/_shame.scss]: Are all the changes in here necessary? They make me a little nervous, since they'll apply to all pages on vets.gov.

**Review by jbalboni — COMMENTED** [review]
If we can get rid of the changes in the shame file, I think this is ready to go. If not, we need to do some more testing.

File: src/sass/_shame.scss
Comment (cehsu) [inline src/sass/_shame.scss]: All of the changes are of the following type:
-alphabetization of properties
-removing leading zeros from decimal values
-removing units from zero values
-trailing whitespace/missing padding
-double colons for pseudo-elements
-using // instead of /**/ for single line comments
-wrapping strings in double quotes
-removing qualifying elements for class selectors for `body .row` and `a.usa-button-primary` within both `.expander-content-inner` and `.expander-content-closed`, as
--it is unlikely that classes will be assigned outside of the `body` element
--`usa-button-primary` is only ever used for anchor tags within `.expander-content-inner` and `.expander-content-closed`
As a result, these will not affect the styling behavior. These were all recommended by the linter along with some other recommendations (for example, using `!important` and id's were also discouraged, but fixing these seemed likely to necessitate additional changes elsewhere). Please let me know if you prefer to revert any of these.

Comment (jbalboni) [inline src/sass/_shame.scss]: @cehsu I think we should remove any changes that affect the specificity of styles, since those could result in visual changes. So changes like removing qualifying elements. The other ones are fine.

**Review by jbalboni — APPROVED** [review]
This can be merged once the specificity related changes are reverted in the shame scss file.

File: src/sass/_shame.scss
Comment (jbalboni) [inline src/sass/_shame.scss]: @cehsu You can make a separate tech debt issue for making the qualifying element changes in that file if you'd like.

**Comment by J-Quag** [comment]
For the link to education about transferring of benefits in the warning message if they select No for Whether the sponsor transferred their benefits:
https://www.dmdc.osd.mil/milconnect/public/faq/Education_Benefits-How_to_Transfer_Benefits

**Comment by cvalarida** [comment]
Turned :point_up: into a ticket [here](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2970).

**Pull request merged**
